src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUContext.java

```diff
@@ -401,21 +406,26 @@ public void cudaFreeHelper(String instructionName, final Pointer toFree) {
 	 */
 	public void cudaFreeHelper(String instructionName, final Pointer toFree, boolean eager) {
 		Pointer dummy = new Pointer();
-		if (toFree == dummy) // trying to free a null pointer
+		if (toFree == dummy) { // trying to free a null pointer
+			if (LOG.isTraceEnabled()) {
+				LOG.trace("GPU : trying to free an empty pointer");
+			}
 			return;
+		}
 		long t0 = 0;
 		if (!cudaBlockSizeMap.containsKey(toFree))
 			throw new RuntimeException(
 					"ERROR : Internal state corrupted, cache block size map is not aware of a block it trying to free up");
 		long size = cudaBlockSizeMap.get(toFree);
 		if (eager) {
 			if (LOG.isTraceEnabled()) {
-				LOG.trace("GPU : eagerly freeing cuda memory [ " + toFree + " ] for instruction " + instructionName
+				LOG.trace("GPU : eagerly freeing cuda memory [ " + toFree + " ] of size " + size + " for instruction " + instructionName
 						+ " on " + this);
 			}
 			if (DMLScript.STATISTICS)
 				t0 = System.nanoTime();
 			cudaFree(toFree);
+			jcuda.runtime.JCuda.cudaDeviceSynchronize();
```

Comment:
`cudaFree` and `cudaMalloc` are synchronous operations. You don't need to put this `cudaDeviceSynchronize` here. (if in fact, that is your intent).